sources/scraping/helpers.py

```diff
@@ -0,0 +1,159 @@
+import threading
+import typing as t
+from typing import Any
+
+import dlt
+import scrapy  # type: ignore
+
+from dlt.common.schema.typing import TWriteDisposition
+from dlt.common.configuration.inject import with_config
+from dlt.common.configuration.specs.base_configuration import (
+    configspec,
+    BaseConfiguration,
+)
+
+from scrapy.crawler import CrawlerRunner  # type: ignore
+from twisted.internet import reactor
+
+from .types import AnyDict
+from .queue import BaseQueue
+from .scrapy.spider import DltSpider
+from .settings import SOURCE_SCRAPY_SETTINGS, SOURCE_SCRAPY_QUEUE_SIZE
+
+
+T = t.TypeVar("T")
+
+
+class Runnable(t.Protocol):
+    def run(self, *args: t.Any, **kwargs: AnyDict) -> Any:
+        pass
+
+
+@configspec
+class ScrapingConfig(BaseConfiguration):
+    # Batch size for scraped items
+    batch_size: int = 20
+
+    # maxsize for queue
+    queue_size: t.Optional[int] = SOURCE_SCRAPY_QUEUE_SIZE
+
+    # result wait timeout for our queue
+    queue_result_timeout: t.Optional[int] = 5
+
+    # List of start urls
+    start_urls: t.Optional[t.List[str]] = None
+
+
+class ScrapyRunner(Runnable):
+    def __init__(
+        self,
+        queue: BaseQueue[T],
+        spider: t.Type[scrapy.Spider],
+        start_urls: t.List[str],
+        settings: AnyDict,
+        include_headers: bool = False,
+    ) -> None:
+        self.queue = queue
+        self.spider = spider
+        self.settings = settings
+        self.start_urls = start_urls
+        self.include_headers = include_headers
+
+    def run(self, *args: Any, **kwargs: AnyDict) -> None:
+        runner = CrawlerRunner(settings=self.settings)
```

Comment:
Question: why use this when you can use (the recommended way) this: https://docs.scrapy.org/en/latest/topics/practices.html#run-scrapy-from-a-script

Reply:
There were some issues with twisted not properly handling teardown process. I will try it again :)

Reply:
Adjusted it. Now we are using `CrawlerProcess`